zerver/webhooks/sonarr/view.py

```diff
@@ -0,0 +1,104 @@
+
+
+@webhook_view("Sonarr")
+@has_request_variables
+def api_sonarr_webhook(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    payload: Dict[str, Any] = REQ(argument_type="body"),
+) -> HttpResponse:
+    body = get_body_for_http_request(payload)
+    subject = get_subject_for_http_request(payload)
+
+    check_send_webhook_message(request, user_profile, subject, body)
+    return json_success()
+
+
+def get_subject_for_http_request(payload: Dict[str, Any]) -> str:
+    if payload["eventType"] != "Test" and payload["eventType"] != "Health":
+        topic = SONARR_TOPIC_TEMPLATE.format(series_title=payload["series"]["title"])
+    elif payload["eventType"] == "Test":
+        topic = SONARR_TOPIC_TEMPLATE_TEST
+    elif payload["eventType"] == "Health":
+        topic = SONARR_TOPIC_TEMPLATE_HEALTH_CHECK.format(level=payload["level"])
+
+    return topic
+
+
+def get_body_for_http_request(payload: Dict[str, Any]) -> str:
+    if payload["eventType"] == "Test":
+        body = get_setup_webhook_message("Sonarr")
+    elif payload["eventType"] == "Health":
+        body = SONARR_MESSAGE_TEMPLATE_HEALTH_CHECK.format(message=payload["message"])
+    elif payload["eventType"] == "Rename":
+        body = SONARR_MESSAGE_TEMPLATE_EPISODES_RENAMED.format(
+            series_title=payload["series"]["title"]
+        )
+    elif payload["eventType"] == "SeriesDelete":
+        body = SONARR_MESSAGE_TEMPLATE_SERIES_DELETED.format(
+            series_title=payload["series"]["title"]
+        )
+    elif payload["eventType"] == "Download" and "isUpgrade" in payload:
+        if payload["isUpgrade"]:
+            data = {
+                "series_title": payload["series"]["title"],
+                "series_number": payload["episodes"][0]["seasonNumber"],
+                "episode_number": payload["episodes"][0]["episodeNumber"],
+                "episode_name": payload["episodes"][0]["title"],
+                "new_quality": payload["episodeFile"]["quality"],
+                "old_quality": payload["deletedFiles"][0]["quality"],
+            }
+
+            body = SONARR_MESSAGE_TEMPLATE_EPISODE_IMPORTED_UPGRADE.format(**data)
+        else:
+            data = {
+                "series_title": payload["series"]["title"],
+                "series_number": payload["episodes"][0]["seasonNumber"],
+                "episode_number": payload["episodes"][0]["episodeNumber"],
+                "episode_name": payload["episodes"][0]["title"],
+            }
+
+            body = SONARR_MESSAGE_TEMPLATE_EPISODE_IMPORTED.format(**data)
+    else:
+        data = {
+            "series_title": payload["series"]["title"],
+            "series_number": payload["episodes"][0]["seasonNumber"],
+            "episode_number": payload["episodes"][0]["episodeNumber"],
+            "episode_name": payload["episodes"][0]["title"],
```

Comment:
The first 4 lines of `data` are the same in all of these so there is potential to make this logic a little simpler. Although this isn't a huge deal! :)

Reply:
@eeshangarg I've actually just done some testing and due to these keys not being available in all of the potential payloads. Its easier to leave as is, as we'd at least have to declare it twice to get round the payloads which don't include it.

Reply:
That makes sense, no worries! Approving! :)